src/hll/estimator.rs

```diff
@@ -0,0 +1,600 @@
+        let k = 1 << lg_config_k;
+        if adj_est > (3 * k) as f64 {
+            return adj_est;
+        }
+
+        // Get linear counting estimate
+        let lin_est = self.get_bitmap_estimate(lg_config_k, cur_min, num_at_cur_min);
+
+        // Blend estimates based on crossover threshold
+        // Use average to reduce bias from threshold comparison
+        let avg_est = (adj_est + lin_est) / 2.0;
+
+        // Crossover thresholds (empirically determined)
+        let crossover = match lg_config_k {
+            4 => 0.718,
+            5 => 0.672,
+            _ => 0.64,
+        };
+
+        let threshold = crossover * (k as f64);
+
+        if avg_est > threshold {
+            adj_est
+        } else {
+            lin_est
+        }
+    }
+
+    /// Get the HIP accumulator value
+    pub fn hip_accum(&self) -> f64 {
+        self.hip_accum
+    }
+
+    /// Get the kxq0 register value
+    pub fn kxq0(&self) -> f64 {
+        self.kxq0
+    }
+
+    /// Get the kxq1 register value
+    pub fn kxq1(&self) -> f64 {
+        self.kxq1
+    }
+
+    /// Check if this estimator is in out-of-order mode
+    pub fn is_out_of_order(&self) -> bool {
+        self.out_of_order
+    }
+
+    /// Set the out-of-order flag
+    ///
+    /// This should be set to true when:
+    /// - Deserializing a sketch from bytes
+    /// - After a merge/union operation
+    pub fn set_out_of_order(&mut self, ooo: bool) {
+        self.out_of_order = ooo;
+        if ooo {
+            // When going out-of-order, invalidate HIP accumulator
+            // (it will be recomputed if needed via composite estimator)
+            self.hip_accum = 0.0;
+        }
+    }
+
+    /// Set the HIP accumulator directly
+    pub fn set_hip_accum(&mut self, value: f64) {
+        self.hip_accum = value;
+    }
+
+    /// Set the kxq0 register directly
+    pub fn set_kxq0(&mut self, value: f64) {
+        self.kxq0 = value;
+    }
+
+    /// Set the kxq1 register directly
+    pub fn set_kxq1(&mut self, value: f64) {
+        self.kxq1 = value;
+    }
+}
+
+/// Compute 1 / 2^value (inverse power of 2)
+#[inline]
+fn inv_pow2(value: u8) -> f64 {
+    if value == 0 {
+        1.0
+    } else if value <= 63 {
+        1.0 / (1u64 << value) as f64
+    } else {
+        f64::exp2(-(value as f64))
+    }
+}
+
+/// Number of standard deviations for confidence bounds
+///
+/// This enum specifies the number of standard deviations to use when computing
+/// upper and lower bounds for cardinality estimates. Higher values provide wider
+/// confidence intervals with greater certainty that the true cardinality falls
+/// within the bounds.
+#[repr(u8)]
+pub enum NumStdDev {
+    /// One standard deviation (\~68% confidence interval)
+    StdDev1 = 1,
+    /// Two standard deviations (\~95% confidence interval)
+    StdDev2 = 2,
+    /// Three standard deviations (\~99.7% confidence interval)
+    StdDev3 = 3,
+}
```

Comment:
```suggestion
pub enum NumStdDev {
    /// One standard deviation (\~68% confidence interval)
    One = 1,
    /// Two standard deviations (\~95% confidence interval)
    Two = 2,
    /// Three standard deviations (\~99.7% confidence interval)
    Three = 3,
}
```

Since we typically use it as `NumStdDev::Variant`, maybe we can use `NumStdDev::One` over `NumStdDev::StdDev1`? This is different from Go, where Go's "enum" is simulated and cannot be qualified with a prefix.

Reply:
I personally don't like spelling out numbers (like 1 vs "One"), but in this context I think your comment makes it nicer :) So why not!

Reply:
Yeah .. `NumStdDev::1` would be better .. But you know it unfortunately can't be an identifier :\